### PR TITLE
asserts: refactor and add support for components in model assertion

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Add support to components in model assertion, which required a small previous refactor of some functions.

### CODE REVIEW

**Review by andrewphelpsj — APPROVED** [review]
Looks good to me!

File: asserts/model.go [hunk 1]
```diff
@@ -283,9 +318,74 @@ func checkModelSnap(snap map[string]interface{}, grade ModelGrade, modelIsClassi
...
+		for key := range compFields {
+			if !strutil.ListContains([]string{"presence", "modes"}, key) {
+				return nil, fmt.Errorf("entry %q of component %s %s is unknown", key, name, what)
+			}
+		}
+		presence, err := checkOptionalStringWhat(compFields, "presence", what)
+		if err != nil {
+			return nil, err
+		}
+		if !strutil.ListContains(validSnapPresences, presence) {
```
Comment (bboozzoo) [inline asserts/model.go:370]: btw. I know the spec doesn't state that, but is there a default value for presence if it's omitted?

File: asserts/model.go [hunk 2]
```diff
@@ -283,9 +318,74 @@ func checkModelSnap(snap map[string]interface{}, grade ModelGrade, modelIsClassi
...
+			return nil, fmt.Errorf("presence of component %q %s must be one of required|optional", name, what)
+		}
+		modes, err := checkStringListInMap(compFields, "modes",
+			fmt.Sprintf("%s %s %s", "modes of component", name, what), validSnapMode)
```
Comment (bboozzoo) [inline asserts/model.go:374]:
```suggestion
			fmt.Sprintf("modes of component %s %s", name, what), validSnapMode)
```

File: asserts/model.go [hunk 3]
```diff
@@ -283,9 +318,74 @@ func checkModelSnap(snap map[string]interface{}, grade ModelGrade, modelIsClassi
...
+		if err != nil {
+			return nil, err
+		}
+		for _, m := range modes {
```
Comment (bboozzoo) [inline asserts/model.go:378]: should we do?
```
if len(modes) == 0 {
    modes = validModes
}
```

**Review by bboozzoo — COMMENTED** [review]
thank you, some questions

File: asserts/model.go (hunk 1 above)
Comment (andrewphelpsj) [reply asserts/model.go:370]: Snaps in the model default to `"required"`, components probably should too for consistency.

File: asserts/model.go (hunk 3 above)
Comment (andrewphelpsj) [reply asserts/model.go:378]: I believe, by default, the component should inherit the modes from its snap.

Comment (alfonsosanchezbeato) [reply asserts/model.go]: Right, I've fixed this now

File: asserts/model.go [hunk 4]
```diff
@@ -283,9 +318,74 @@ func checkModelSnap(snap map[string]interface{}, grade ModelGrade, modelIsClassi
...
 	}, nil
 }
 
+func checkComponentsForMaps(m map[string]interface{}, validModes []string, what string) (map[string]ModelComponent, error) {
+	const compsField = "components"
+	value, ok := m[compsField]
+	if !ok {
+		return nil, nil
+	}
+	comps, ok := value.(map[string]interface{})
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply asserts/model.go:347]: Done (was a bit tricky though because of `go fmt`)

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the reviews, I've addressed the comments now

**Review by bboozzoo — APPROVED** [review]
LGTM

**Comment by alfonsosanchezbeato** [comment]
Commits squashed now

**Pull request merged**